arangod/RocksDBEngine/RocksDBCollection.cpp

```diff
@@ -69,16 +73,499 @@
+
+  arangodb::velocypack::Builder builder;
+  std::pair<std::uint64_t, std::uint64_t> outputPair;
+  ::IndicesMap indicesMap;
+  {
+    arangodb::velocypack::ObjectBuilder collectionObjectGuard(&builder);
+    std::uint64_t objectId =
+        VelocyPackHelper::stringUInt64(oldProps, StaticStrings::ObjectId);
+    std::uint64_t tempObjectId =
+        VelocyPackHelper::stringUInt64(oldProps, StaticStrings::TempObjectId);
+    outputPair = idFunc(objectId, tempObjectId);
+    for (auto pair : arangodb::velocypack::ObjectIterator(oldProps)) {
+      ::handlePropertiesEntryForObjectIdUpdate(collection, builder, idFunc, setUpgradedProperties,
+                                               indicesMap, pair.key, pair.value);
+    }
+    builder.add(StaticStrings::ObjectId,
+                arangodb::velocypack::Value(std::to_string(outputPair.first)));
+    builder.add(StaticStrings::TempObjectId,
+                arangodb::velocypack::Value(std::to_string(outputPair.second)));
+  }
+  auto value = arangodb::RocksDBValue::Collection(builder.slice());
+  batch.Put(arangodb::RocksDBColumnFamily::definitions(), key.string(), value.string());
+  res = arangodb::rocksutils::convertStatus(db.Write(wo, &batch));
+  if (res.fail()) {
+    return res;
+  }
+  auto cleanup = arangodb::scopeGuard([oldProps, &db, &key]() -> void {
+    rocksdb::WriteOptions wo;
+    rocksdb::WriteBatch batch;
+    auto value = arangodb::RocksDBValue::Collection(oldProps);
+    batch.Put(arangodb::RocksDBColumnFamily::definitions(), key.string(), value.string());
+    db.Write(wo, &batch);
+  });
+
+  res = ::setObjectIdsForCollection(collection, outputPair, indicesMap);
+  if (res.fail()) {
+    return res;
+  }
+
+  cleanup.cancel();  // succeeded, no cleanup needed
+
+  return res;
+}
+
+arangodb::Result commitBatch(rocksdb::WriteOptions& wo,
+                             rocksdb::WriteBatch& batch, rocksdb::DB& db) {
+  rocksdb::Status r = db.Write(wo, &batch);
+  arangodb::Result res = arangodb::rocksutils::convertStatus(r);
+  if (res.ok()) {
+    batch.Clear();
+  }
+  return res;
+}
+
+void rewriteDocument(rocksdb::WriteBatch& batch, arangodb::RocksDBKey& key,
+                     arangodb::RocksDBKeyBounds& bounds,
+                     rocksdb::Slice oldValue, std::uint64_t tempObjectId) {
+  arangodb::velocypack::Slice oldDocument = arangodb::RocksDBValue::data(oldValue);
+  arangodb::velocypack::Builder builder;
+  arangodb::velocypack::Slice revSlice =
+      oldDocument.get(arangodb::StaticStrings::RevString);
+  TRI_ASSERT(revSlice.isString());
+  arangodb::velocypack::ValueLength l;
+  char const* p = revSlice.getString(l);
+  TRI_voc_rid_t revision = TRI_StringToRid(p, l, false);
+  {
+    arangodb::velocypack::ObjectBuilder guard(&builder);
+    for (auto pair : arangodb::velocypack::ObjectIterator(oldDocument)) {
+      builder.add(pair.key);
+      builder.add(pair.value);
+    }
+  }
+
+  arangodb::velocypack::Slice newDocument = builder.slice();
```

Comment:
We need to make sure here (and I am not sure if this is currently guaranteed) that the new documents that are created have their system attributes at the beginning.
Documents in the database must have `_key`, `_id` and `_rev` first, followed by `_from` and `_to` for edge collections. All other user-defined attributes must go afterwards.

I also don't understand why we do iterate over the old document's attributes just to store them unmodified in a new Builder. If we don't rewrite the document data, then the contents of `newDocument` and `oldDocument` should be identical. And we can then remove the whole iterate-over-all-attributes thing and reuse `oldDocument` as it is. This will also eliminate my concerns.

Reply:
Remnant from a temporary state. Can just copy the document as suggested.